### PR TITLE
First `Deframer` implementation

### LINKED ISSUE #24: Deframer

### PR DESCRIPTION
**Pull request by juanimedone**
Closes #24.

This PR introduces an HDLC deframer implementation with robust frame extraction logic, handling garbage, partial flags (they come in chunks), and buffer overflows. It includes:

- The `Deframer` struct and state machine for reading HDLC frames from a bit stream.
- Extraction of frames between flag sequences, with buffer management and overflow protection. A `MAX_BUFFER_LEN` is set.
- A comprehensive test suite covering:
  - Single and multiple frames.
  - Frames with/without garbage.
  - Edge cases (missing flags, buffer limits, split flags).
  - Buffer overflow and recovery scenarios.

### Next steps
- #35 (#37)
- #42 
- #43

### CODE REVIEW

File: hdlc/src/deframer.rs [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+    }
+
+    // the function return is temporary for testing purposes
+    pub fn run(&mut self) -> Vec<RawFrame> {
+        let mut frames = Vec::new();
+        while let Ok(new_bits) = self.reader.recv() {
+            // for testing purposes
+            if new_bits.is_empty() {
+                break;
+            }
```
Comment (lazcanoluca) [inline hdlc/src/deframer.rs:41]: I think it should be possible to receive an empty `Vec`. Unusual, yes, but I don't think it should stop the deframer.

File: hdlc/src/deframer.rs [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+                self.buffer.push_back(bit);
+            }
+
+            let new_frames = self.find_frames();
+            frames.extend(new_frames);
+            // let packets = frames
+            //     .into_iter()
+            //     .map(|frame| Packet::new(frame))
+            //     .collect::<Vec<Packet>>();
+            // send these packets to the web app?
```
Comment (lazcanoluca) [inline hdlc/src/deframer.rs:53]: We could have a "writer", a channel maybe, where we stream the frames to. It would ease testing and solve the "sender part" of the Deframer's API.

File: hdlc/src/deframer.rs [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+// 4096 bits (512 bytes) is a safe upper bound for most use cases
+const MAX_BUFFER_LEN: usize = 4096;
+const FLAG_ARRAY: [bool; 8] = [false, true, true, true, true, true, true, false];
+
+pub enum ParserState {
+    SearchingStartSync,
+    SearchingEndSync,
+}
+
+struct Deframer {
```
Comment (lazcanoluca) [inline hdlc/src/deframer.rs:14]:
```suggestion
pub struct Deframer {
```

File: hdlc/src/deframer.rs [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+            }
+
+            let new_frames = self.find_frames();
+            frames.extend(new_frames);
+            // let packets = frames
+            //     .into_iter()
+            //     .map(|frame| Packet::new(frame))
+            //     .collect::<Vec<Packet>>();
+            // send these packets to the web app?
+            // Packet should be an interface so multiple packet types can be implemented
```
Comment (lazcanoluca) [inline hdlc/src/deframer.rs:54]: Good idea, Packet could be a trait and when creating the Deframer we pass it a PacketBuilder that will dynamically create the desired packet type. We could create an issue for this, for later.

**Review by alendavies — APPROVED** [review]

File: hdlc/src/deframer.rs [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+            if slice == FLAG_ARRAY {
+                match self.state {
+                    ParserState::SearchingStartSync => {
+                        self.state = ParserState::SearchingEndSync;
+                        // update the index, so it starts looking for the end sync
+                        self.idx += 8;
+                    }
+                    ParserState::SearchingEndSync => {
+                        // drain the whole frame between syncs
+                        let frame_bits = self.buffer.drain_range(0, self.idx + 8);
```
[the remark this answers is not in the document]
Comment (juanimedone) [reply hdlc/src/deframer.rs:84]: Yes, as for now I left it like this because `Frame::new` is needed to be able to change `find_frames` logic and allow reading two frames separated by a single SYNC. 

`ParserState::SearchingEndSync` would have no reason to exist, as any SYNC (except the first one read) could be a start and end SYNC at the same time. The final implementation should take all the bits read between two SYNCs and try to create a new `Frame` with that content. If this fails, it's considered as garbage and discarded.

I just created an issue (#45) for this, to be addressed later.

File: hdlc/src/bitvecdeque.rs [hunk 6]
```diff
@@ -0,0 +1,673 @@
...
+    pub fn from_bits<I: IntoIterator<Item = bool>>(bits: I) -> Self {
+        let mut deque = Self::new();
+        for bit in bits {
+            deque.push_back(bit);
+        }
+        deque
+    }
+
+    // Método para crear desde BitVec
+    pub fn from_bitvec(bitvec: BitVec) -> Self {
```
[the remark this answers is not in the document]
Comment (juanimedone) [reply hdlc/src/bitvecdeque.rs]: Done.

File: hdlc/src/deframer.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lazcanoluca) [reply hdlc/src/deframer.rs:84]: Awesome!

File: hdlc/src/deframer.rs (hunk 1 above)
Comment (juanimedone) [reply hdlc/src/deframer.rs]: Ok, I just changed it. I made it that way before only for testing purposes, but found another way to stop the reader in the tests.

File: hdlc/src/deframer.rs (hunk 4 above)
Comment (juanimedone) [reply hdlc/src/deframer.rs:53]: Yes, I just included it in #46.

File: hdlc/src/deframer.rs [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+        let mut frames = Vec::new();
+
+        while self.buffer.len() - self.idx >= 8 {
+            // Drop buffer if it grows too large (prevents DoS via never-ending garbage)
+            if self.buffer.len() > MAX_BUFFER_LEN {
+                self.buffer.clear();
+                self.idx = 0;
+                self.state = ParserState::SearchingStartSync;
+                break;
+            }
```
[the remark this answers is not in the document]
Comment (juanimedone) [reply hdlc/src/deframer.rs:68]: You mean instantiating a new `Deframer` or a new `buffer` when `MAX_BUFFER_LEN` is reached? Otherwise I don't see how this functionality could be implemented in `Deframer::new`.

File: hdlc/src/deframer.rs (hunk 2 above)
Comment (lazcanoluca) [reply hdlc/src/deframer.rs:52]: > that writer should send packets

True!

File: hdlc/src/deframer.rs [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+        let mut frames = Vec::new();
+        while let Ok(new_bits) = self.reader.recv() {
+            // for testing purposes
+            if new_bits.is_empty() {
+                break;
+            }
+
+            for bit in new_bits {
+                self.buffer.push_back(bit);
+            }
```
[the remark this answers is not in the document]
Comment (lazcanoluca) [reply hdlc/src/deframer.rs:44]: Oh I see! Makes sense :)

File: hdlc/src/deframer.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (lazcanoluca) [reply hdlc/src/deframer.rs:68]: Resolved in https://github.com/AstarAeroespacial/rustar-gs/pull/44/files/24b7126c7b9f317478bae98af9e861a1b86fbfa2#r2205849224.

**Review by lazcanoluca — APPROVED** [review]
Awesome!
